### PR TITLE
Operations

### PR DESCRIPTION
**Pull request by noakrams**

### CODE REVIEW

File: calc.s [hunk 1]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    pop esi                               ; restore pointer
+    mov ecx, 0
+    mov cl, byte[esi]                     ; get number that needs to be copied
+    mov byte[eax + NUMBER], cl            ; save number in the malloced node
+    mov ebx, 0
+    mov bl, byte[OSC]
+    mov edi, dword[OSS]
+    mov [edi + 4*ebx], eax                ; save first node in place
+    cmp dword[esi + NEXT], 0              ; if we dont have nodes to copy anymore so we are done
+    mov edi, [edi + 4*ebx]                ; edi points to first node that has been allocated
```
Comment (IdanTo1) [inline calc.s:725]: why is it between the cmp and jmp?

File: calc.s [hunk 2]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    push esi
+    push dword 5
+    call malloc
+    add esp, 4                            ; balance stack
+    pop esi                               ; restore registers
+    pop edi
+    mov ecx, 0
+    mov cl, byte[esi]                     ; save number that needs to be copied
+    mov byte[eax + NUMBER], cl            ; copy number to allocated node
+    mov dword[edi+NEXT], eax              ; initialize previous node next to point to the new node
```
Comment (IdanTo1) [inline calc.s:743]: move it a bit up, to be right after the pop. for context

**Comment by IdanTo1** [comment]
+
1. add debug support (should be pretty easy for you)
2. add inc dword[op_ctr] *before* every method

File: calc.s [hunk 3]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    mov ecx, 0xF                          ; move mask 00001111 to cl
+    and cl, bl                            ; get the second digit to cl
+    shr bl, 4                             ; get first digit to bl
+    pushad                                ; save registers before printf
+    pushfd
+    push ebx                              ; push first digit to printf
+    push print_format                     ; push format to printf
+    call printf
+    add esp, 4                            ; restore stack after pushing pointer of format
+    pop ebx                               ; restore ebx to save stack balance
```
[the remark this answers is not in the document]
Comment (noakrams) [reply calc.s:311]: You have a good point

File: calc.s [hunk 4]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    pushad                                ; same proccess here as above with the second digit
+    pushfd
+    push ecx
+    push print_format
+    call printf
+    add esp, 4
+    pop ebx
+    popfd
+    popad
+    cmp eax, 0                            ; check if the printed byte was the frist one on the list
```
[the remark this answers is not in the document]
Comment (noakrams) [reply calc.s:324]: oops

File: calc.s [hunk 5]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    push ebx                              ; push byte number (two digits) to printf
+    push print_format                     ; push format pointer
+    call printf
+    add esp, 4                            ; balance stack
+    pop ebx                               ; balance stack
+    popfd                                 ; restore flags and registers
+    popad
+    cmp eax, 0                            ; check if the printed byte was the the only one on the list
+    je check_if_debug                     ; if so, check if debug
+    jmp print_reverse                     ; else, return to print_reverse and print the next byte of the operand
```
[the remark this answers is not in the document]
Comment (noakrams) [reply calc.s:363]: I explained in row 307 when I called the label

File: calc.s [hunk 6]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    jmp push_counter
+    two_digits:
+    add ecx, ecx                          ; each link has two digits
 
-mult:
+    push_counter:
+    mov edi, [esi + 4*eax]                ; edi points to the head of the operand
+    mov ebx, 0
+    mov ebx, 0xFF                         ; store 0b11111111 to bl
+    and ebx, ecx
```
[the remark this answers is not in the document]
Comment (noakrams) [reply calc.s:571]: If ecx is bigger than FF? I saw in the forum that they said it is possible

File: calc.s (hunk 1 above)
Comment (noakrams) [reply calc.s:725]: oops

File: calc.s [hunk 7]
```diff
@@ -272,19 +274,488 @@ strtol16:
...
+    mov dword[edi + NEXT], 0              ; initialize next to null
+    jmp myCalc
+
+    free_the_zeros:
+    cmp edx, 1                            ; check if the result is zero
+    je res_is_zero                        ; if so, jump here
+    mov edi, dword[edx + NEXT]            ; else, get the first zero node that need to be freed
+    mov dword[edx + NEXT], 0              ; initialize next of last node with null 
+    push edi                              ; push node zeros
+    call free_operand                     ; free the zeros
```
[the remark this answers is not in the document]
Comment (IdanTo1) [reply calc.s:466]: oh, my bad. you're releasing edi.. all good

File: calc.s [hunk 8]
```diff
@@ -487,11 +494,21 @@ bitwise_AND:
     push ecx
     call free_operand                     ; free this operand
     add esp, 4
-    jmp myCalc                            ; return to main loop
+    jmp return_and                        ; return to main loop
```
Comment (IdanTo1) [inline calc.s:498]: why? your're here (:

File: calc.s (hunk 6 above)
[the remark this answers is not in the document]
Comment (IdanTo1) [reply calc.s:605]: ohhh ok, nice!

**Pull request merged**